src/web/core/src/main/java/org/geoserver/web/wicket/browser/FileRootsFinder.java

```diff
@@ -0,0 +1,176 @@
+ * utility to match file paths in said roots
+ */
+public class FileRootsFinder implements Serializable {
+    
+    /**
+     * Utility so split and rebuild paths accounting for ResourceStore little own illusion
+     * of working on a *nix file system regardless of the actual file system
+     */
+    static class PathSplitter {
+
+        String separator;
+        boolean dataDirectoryPath;
+        String base;
+        String name;
+        
+        public PathSplitter(String input, boolean dataDirectoryPath) {
+            // decide which separator to use based on data dir vs actual file system
+            this.separator = dataDirectoryPath ? "/" : File.separator;
+            this.dataDirectoryPath = dataDirectoryPath;
+            
+            // remove protocol part if needed (
+            if (input.startsWith("file:")) {
+                if (input.startsWith("file:/")) {
+                    input = input.substring(6);
+                } else {
+                    input = input.substring(5);
+                }
+            }
+
+            // split input into base and last segment
+            int idx = input.lastIndexOf(separator);
+            if (idx >= 0) {
+                base = input.substring(0, idx);
+                name = input.substring(idx + 1, input.length()).toLowerCase();
+            } else {
+                base = separator;
+                name = input.toLowerCase();
+            }
+
+            // fix base in case of data dir
+            if (dataDirectoryPath) {
+               base = Paths.convert(base); 
+            }
+        }
+        
+        private String buildPath(String name) {
+            // Data dir relative path weirdness, the protocol has to be 
+            // file:/ instead of file:// or it won't work. 
+            String prefix = dataDirectoryPath ? "file:" : "file://";
+            // make data dir relative paths actually relative despite user's input
+            String localBase = base;
+            if(dataDirectoryPath && localBase.startsWith(separator)) {
+                localBase = base.substring(1);
+            }
+            if (localBase.endsWith(separator)) {
+                return prefix + localBase + name;
+            } else {
+                return prefix + localBase + separator + name;
+            }
+        }
+
+    }
+
+    private ArrayList<File> roots;
+    private File dataDirectory;
+
+    public FileRootsFinder(boolean includeDataDir) {
+        this(GeoServerFileChooser.HIDE_FS, includeDataDir);
+    }
+    
+    public FileRootsFinder(boolean hideFileSystem, boolean includeDataDir) {
+        // build the roots
+        roots = new ArrayList<File>();
+        if (!hideFileSystem) {
+            roots.addAll(Arrays.asList(File.listRoots()));
+        }
+        Collections.sort(roots);
+
+        // TODO: find a better way to deal with the data dir
+        GeoServerResourceLoader loader = getLoader();
+        dataDirectory = loader.getBaseDirectory();
```

Comment:
Q: What would a better way look like? That class has the responsibility of knowing the base directory location, which is better than half a dozen classes as previous ...

Reply:
I've already spent way more spare time than I wanted on this one, if you care about changing it make a pull request yourself. If you think this is not acceptable I'll just close the pull request and forget about it.

Reply:
Andrea no need to be harsh, I was asking for feedback on the API, I am not aware of a better way to get the base directory location, am I missing something?